node/primitives/src/approval.rs

```diff
@@ -77,6 +221,33 @@ pub enum AssignmentCertKind {
 	},
 }
 
+/// Certificate is changed compared to `AssignmentCertKind`:
+/// - introduced RelayVRFModuloCompact
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub enum AssignmentCertKindV2 {
+	/// An assignment story based on the VRF that authorized the relay-chain block where the
+	/// candidate was included combined with a sample number.
+	///
+	/// The context used to produce bytes is [`v2::RELAY_VRF_MODULO_CONTEXT`]
+	RelayVRFModulo {
+		/// The sample number used in this cert.
+		sample: u32,
+	},
+	/// Multiple assignment stories based on the VRF that authorized the relay-chain block where the
+	/// candidates were included.
+	///
+	/// The context is [`v2::RELAY_VRF_MODULO_CONTEXT`]
+	RelayVRFModuloCompact,
```

Comment:
I'd expect `RelayVRFModuloCompact` gains a `claimed_cores: Vec<u32>` field, but which goes into the extra data, not the VRF input.

Alternatively `AssignmentCertV2` could gain a `claimed_cores: Vec<u32>` field, which gets hashed into extra, but then you should handle this field in the delay context somehow, either make it be correct, or make it be empty.

In my original draft, this "story" data could be pushed either into the extra data or into the VRF input, but I only used it for randomness there. https://github.com/paritytech/polkadot/pull/1558/files

Reply:
Very odd.  I maybe miss-read something then.  Where is the other place where `core_indices` exists?

I see `check_assignment_cert` takes `core_indices` separately, when it should use the `core_indices` in `AssignmentCertV2`.  I guess this was inherited from before? 

It's odd I'd say "since it's uniquely determined" though, maybe I thought you were not yet doing subsets, or maybe I though `sample` was really `highest_sample`, but neither seems likely.  Very odd.

Reply:
I guess we can remove the `core_indices` argument to `check_assignment_cert` and change the outer envelop?

Reply:
We can't change the outer envelope, the candidate indices in there are used for book keeping of assignments without checking signature in `approval-distribution`.

Reply:
Wouldn't it be fine to extract the inner value for bookkeeping and then handle a mismatch the same way as is currently done?

Reply:
Why can't we just change the outer envelop?  Actually, why can't we just remove the outer envelop entirely?  It sounds risky to be passing unsigned messages.

Reply:
Ok, I'll look into removing the candidate indices and use the inner cert core indices.

Reply:
I started working on this and it is doable @rphmeier . However, this would add some serious additional code complexity since we need to also continue to support the v1 assignments for a while. I suggest to handle this in https://github.com/paritytech/polkadot/issues/6988 as this PR is already too big.

WDYT ?